runtime/src/iree-amd-aie/aie_runtime/amsel_generator.cc

```diff
@@ -0,0 +1,230 @@
+// Copyright 2024 The IREE Authors
+//
+// Licensed under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+
+#include "amsel_generator.h"
+
+#include <deque>
+#include <limits>
+
+#define DEBUG_TYPE "iree-aie-runtime-amsel-generator"
+
+namespace mlir::iree_compiler::AMDAIE {
+
+LogicalResult MSelGenerator::addConnection(
+    const PhysPortAndID &src, const SmallVector<PhysPortAndID> &dsts) {
+  SmallVector<Port> dstPorts =
+      llvm::map_to_vector(dsts, [](const PhysPortAndID &physPortAndID) {
+        return physPortAndID.physPort.port;
+      });
+  if (!connectionToMSel.contains(dstPorts)) {
+    if (curMSel >= numMSels) {
+      LLVM_DEBUG(llvm::dbgs()
+                 << "Trying to assign msel (" << std::to_string(curMSel)
+                 << ") larger than number of msels available ("
+                 << std::to_string(numMSels) << ") for " << src << "\n");
+      return failure();
+    }
+    connectionToMSel[dstPorts] = curMSel++;
+  }
+  physPortToConnection[src] = dstPorts;
+  srcPhysPortAndIDs.push_back(src);
+  return success();
+}
+
+LogicalResult MSelGenerator::addConnections(
+    const SmallVector<PhysPortAndID> &srcs,
+    const SmallVector<SmallVector<PhysPortAndID>> &dsts) {
+  for (auto &&[src, dstsElem] : llvm::zip(srcs, dsts)) {
+    if (failed(addConnection(src, dstsElem))) return failure();
+  }
+  return success();
+}
+
+uint8_t MSelGenerator::getNumMSelsForConnections(
+    const SmallVector<PhysPortAndID> &srcs,
+    const SmallVector<SmallVector<PhysPortAndID>> &dsts) const {
+  uint8_t numMSelsNeeded{0};
+  for (const SmallVector<PhysPortAndID> &dstsElem : dsts) {
+    assert(numMSelsNeeded < std::numeric_limits<uint8_t>::max());
```

Comment:
um isn't this impossible since the type is `uint8_t`?

Reply:
It's possible because `std::numeric_limits<uint8_t>::max()` is the max uint8_t, so 255. I check it  here before it would overflow.